Plugin/Magento/CatalogInventory/Api/StockItemRepositoryPlugin.php

```diff
@@ -0,0 +1,169 @@
+    private $logger;
+
+    /**
+     * @param ProductEventManagerInterface $productEventManager
+     * @param ProductEventInterfaceFactory $productEventFactory
+     * @param RunInstantProductEvent $runInstantProductEvent
+     * @param ProductRepositoryInterface $productRepository
+     * @param Config $config
+     * @param ProductWebsiteLink $productWebsiteLink
+     * @param Logger $logger
+     */
+    public function __construct(
+        ProductEventManagerInterface $productEventManager,
+        ProductEventInterfaceFactory $productEventFactory,
+        RunInstantProductEvent $runInstantProductEvent,
+        ProductRepositoryInterface $productRepository,
+        Config $config,
+        ProductWebsiteLink $productWebsiteLink,
+        Logger $logger
+    ) {
+        $this->productEventManager = $productEventManager;
+        $this->productEventFactory = $productEventFactory;
+        $this->runInstantProductEvent = $runInstantProductEvent;
+        $this->productRepository = $productRepository;
+        $this->config = $config;
+        $this->productWebsiteLink = $productWebsiteLink;
+        $this->logger = $logger;
+    }
+
+    /**
+     * Publish product event after stock item save process
+     *
+     * @param StockItemRepositoryInterface $subject
+     * @param callable $proceed
+     * @param StockItemInterface $stockItem
+     * @return StockItemInterface
+     */
+    public function aroundSave(
+        StockItemRepositoryInterface $subject,
+        callable $proceed,
+        StockItemInterface $stockItem
+    ): StockItemInterface {
+        $oldStockItem = ($stockItem->getItemId()) ?
+            $subject->get($stockItem->getItemId()) : null;
+
+        /** @var StockItemInterface $stockItemUpdated */
+        $stockItemUpdated = $proceed($stockItem);
+        $websiteIds = $this->productWebsiteLink->getWebsiteIdsByProductId($stockItem->getProductId());
+        foreach ($websiteIds as $websiteId) {
+            if (!$this->config->getIsCatalogIngestionEnabled($websiteId)) {
+                continue;
+            }
+            $this->processStockItem($oldStockItem, $stockItemUpdated, $websiteId);
+        }
+        return $stockItemUpdated;
+    }
+
+    /**
+     * Process stock item product events
+     *
+     * @param StockItemInterface $stockItemUpdated
+     * @param StockItemInterface|null $oldStockItem
+     * @param int|null $websiteId
+     * @return void
+     */
+    private function processStockItem(
+        StockItemInterface $stockItemUpdated,
+        StockItemInterface $oldStockItem = null,
+        int $websiteId = null
+    ): void
+    {
+        try {
+            if (!$oldStockItem ||
+                $oldStockItem->getQty() != $stockItemUpdated->getQty() ||
+                $oldStockItem->getIsInStock() != $stockItemUpdated->getIsInStock() ||
+                $oldStockItem->getManageStock() != $stockItemUpdated->getManageStock()
+            ) {
+                if ($oldStockItem &&
```

Comment:
I am personally a big fun of moving catalogInventory business logic out from all plugins and events.
So the purpose of the plugin: identify if a product was created / updated and pass all necessary data to Catalog ingestion command.
Catalog ingestion command check settings and do actions depend on those settings.

We can see that we properly splitter the code id we don't check catalog ingestion settings in plugins/events.
We can of course check if CatalogIngestion enabled but not other settings.

In this specific case we can move out code from line 147 to 163, it contains just a few parameter:
- websiteID
- productID
- type: update or create
- flag about stock status changes (null for create case)

Let's move business logic in lin

Reply:
Moved to new class: Model/CatalogIngestion/ProductEventProcessor.php

https://github.com/BoltApp/bolt-magento2/pull/1560/files#diff-5034980d5cd6db4c4e94a8806840ed03c1ffb376c08b2c6c279e12e7dd77b3d2